File: BinarySearchTree/BST.py
# Binary Search Tree implementation

class Node:
    def __init__(self, data=None) -> None:
        self.data = data
        self.right = None
        self.left = None


class BST:
    def __init__(self, data=None):
        if data:
            self.root = Node(data)
            self.size = 1
        else:
            self.root = Node()
            self.size = 0
    
    def _insert(self, node: Node, data):
        if node is None:
            return Node(data)
        
        if(data > node.data):
            node.right = self._insert(node.right, data=data)

        if(data < node.data):
            node.left = self._insert(node.left, data=data)

        return node 
    
    def _remove(self, node: Node, data):
        if node is None:
            return
        
        if data > node.data:
            node.right = self._remove(node.right, data)
        elif data < node.data:
            node.left = self._remove(node.left, data)
        else:
            # Here the node is found!
            if node.left is None:
                temp = node.right or None
                del node
                return temp
            elif node.right is None:
                temp = node.left
                del node
                return temp
            else:
                # We need to find successor!
                succ: Node = node.right

                while succ.left is not None:
                    succ = succ.left
                
                node.data = succ.data
                self._remove(node.right, succ.data)
                return node
                
        # This return is essential as it makes the node structure stable
        return node

    def _contains(self, node: Node, data):
        if node == None:
            return False

        if node.data == data:
            return True
        
        if data > node.data:
            return self._contains(node.right, data)
        if data < node.data:
            return self._contains(node.left, data)
        
    def _preOrderTraversal(self, node: Node):
        if(node  == None):
            return
        print(node.data)
        self._preOrderTraversal(node.left)
        self._preOrderTraversal(node.right)

    def contains(self, data):
        if data is None:
            return ValueError("Invalid Data!")

        return self._contains(self.root, data)

    def getRoot(self):
        return self.root.data

    def insert(self, data):
        if data is None:
            return ValueError("Data must be valid!")
        else:
            self.root = self._insert(self.root, data=data)
            self.size = self.size + 1
            return 
        
    def remove(self, data):
        if data == None:
            return ValueError("Invalid Data.")
        elif self.size == 0:
            return IndexError("Tree is empty!")
        else:
            if self.contains(data) == True:
                self._remove(self.root, data)
                self.size = self.size - 1
                return 
            else:
                return KeyError("Key doesn't exist!")

    def preOrderTraversal(self):
        if self.root is None:
            print("Tree is empty!")
            return;
        self._preOrderTraversal(self.root)


tree = BST(12)
tree.insert(11)
tree.insert(1)
tree.insert(11)
tree.insert(20)
tree.remove(100)
tree.preOrderTraversal()
    

File: cpp-implementation/BinarySearchTree/BST.cpp
#include <iostream>

using namespace std;


template <typename T> struct Node {
    T data;
    Node* left;
    Node* right;

    Node(T d)
    {
        this->data = d;
        this->left = nullptr;
        this->right = nullptr;
    }
};


template <typename T> class BinarySearchTree
{
    private:
        Node<T>* root;
        int size;
    
    public:
        BinarySearchTree()
        {
            this->root = nullptr;
            this->size = 0;
        }

        BinarySearchTree(T data)
        {
            Node<T>* new_node = new Node<T>(data);
            this->root = new_node;
            this->size = 1;
        }

        Node<T>* get_root() {return this->root; }

        void add(T data)
        {
            Node<T>* new_node = new Node<T>(data);
            if(this->root == nullptr)
            {
                root = new_node;
                size++;
                return;
            }

            Node<T>* node = this->root;
            Node<T>* parent = nullptr;

            while(node != nullptr){
                if(data > node->data){
                    parent = node;
                    node = node->right;
                }else {
                    parent = node;
                    node = node->left;                
                }
            }

            if(data > parent->data)
                parent->right = new_node;
            else
                parent->left = new_node;
            
            size++;
            return;
        }

        Node<T>* remove(Node<T>* node, T data)
        {
            if(node == nullptr)
                return node;
           
            if(data > node->data)
                node->right = remove(node->right, data);  
            else if(data < node->data)
                node->left = remove(node->left, data);
            else
            {
                if(node->left == nullptr){
                    Node<T>* temp = node->right;
                    delete node;
                    return temp;
                }
                else if(node->right == nullptr){
                    Node<T>* temp = node->left;
                    delete node;
                    return temp;
                }
                if(node->left && node->right)
                {
                    // Let's find rightmost node;
                    Node<T>* rightmost = node->right;

                    while(rightmost && rightmost->left)
                        rightmost = rightmost->left;
                    
                    node->data = rightmost->data;
                    node->right = remove(node->right, rightmost->data);
                } 
            }
            return node;
        }

        Node<T>* contains(Node<T>* node, T data)
        {
            if(node)
            {
                if(node->data == data)
                    return node;
                
                if(data > node->data)
                    return find(node->right, data);
                else   
                    return find(node->left, data);
            }
            else
                return nullptr;
        }

        void in_order_traversal(Node<T>* node)
        {
            if(node == nullptr) return;
            in_order_traversal(node->left);
            cout << node->data << " ",
            in_order_traversal(node->right);
        }
};


int main (int argc, char * argv[])
{
    BinarySearchTree<int> tree(12);
    Node<int>* root = tree.get_root();

    tree.add(5);
    tree.add(15);
    tree.add(13);
    tree.add(1);

    tree.remove(root, 12);

    tree.in_order_traversal(root);
    return 0;
}